rust/arrow/src/array/array_primitive.rs

```diff
@@ -94,6 +94,32 @@ impl<T: ArrowPrimitiveType> PrimitiveArray<T> {
         let offset = i + self.offset();
         unsafe { *self.raw_values.as_ptr().add(offset) }
     }
+
+    /// Creates a PrimitiveArray based on an iterator of values without nulls
+    pub fn from_iter_values<I: IntoIterator<Item = T::Native>>(iter: I) -> Self {
+        let iter = iter.into_iter();
+        let (_, data_len) = iter.size_hint();
+        let data_len = data_len.expect("Iterator must be sized"); // panic if no upper bound.
+
+        let mut val_buf = MutableBuffer::new(
+            data_len * mem::size_of::<<T as ArrowPrimitiveType>::Native>(),
+        );
+
+        iter.for_each(|item| {
+            val_buf.push(item);
+        });
```

Comment:
Could we wait for #9235? It introduces a method to extend a `MutableBuffer` out of an iterator of Native types. It will also allow to drop the `.expect("Iterator must be sized")`, since it will be possible to extend it out of unsized iterators.

Reply:
Yes, sure 👍  sounds like a good idea

Reply:
#9235 is merged

Reply:
@jorgecarleitao is the new usage with `extend` what you meant?

We also have to know the final count for creating the array, are you thinking of calculating that with `val_buf.len() / mem::size_of::<<T as ArrowPrimitiveType>::Native>()` or is there an other way?

Reply:
Yes, that is what I meant: it should now be possible to do

```rust
let values = vec![0i32, 1];

let values = values.iter().map(|x| x + 1);

let buffer: Buffer: values.collect();
```

wrt to the `len`: that is a good point that I have had not thought about. 👍 

One option is to do what you wrote. Another could be (I haven't tried to compile this, as `Fn` could become `FnMut`):

```
let mut count = 0;
let iter = iter.map(|x| {
   count += 1;
x
});
```

I would probably have taken your idea, though :)

Reply:
Ah with just `Buffer` and `collect` this is even more clean. I adapted the code with the new API.